source/user/sgx.rst

```diff
@@ -27,27 +27,54 @@ Any :attr:`scf.hf.SCF` object :attr:`mf` can be converted to an equivalent objec
+The P-junction screening threshold is determined by :attr:`mf.direct_scf_tol`.
+
+Direct evaluation of the J matrix can be used by setting :attr:`mf.with_df.direct_j = True`, but this is much slower than SGX-J or DF-J and only intended for testing purposes.
+
 Adjustable Parameters
 =====================
 
-Calling the :attr:`sgx_fit` function on an :attr:`scf.hf.SCF` object returns an equivalent :attr:`SGXHF` object with a :attr:`with_df` attribute that handles SGX integration. To use a non-default auxiliary basis (for :attr:`dfj=True`), :attr:`auxbasis` can be specified in the call to :attr:`sgx_fit`. In addition, there are five main adjustable parameters for this object:
+Calling the :attr:`sgx_fit` function on an :attr:`scf.hf.SCF` object returns an equivalent :attr:`SGXHF` object with a :attr:`with_df` attribute that handles SGX integration. To use a non-default auxiliary basis (for :attr:`dfj=True`), :attr:`auxbasis` can be specified in the call to :attr:`sgx_fit`. In addition, there are various adjustable parameters for SGX:
+
+:attr:`with_df` attributes:
 
 * :attr:`grids_level_i`: The grid level to use for initial SCF iterations.
 * :attr:`grids_level_f`: The grid level to use for final SCF iterations.
 * :attr:`grids_thrd`: The grid points where no atomic orbital is significant (has a value greater than this threshold) are removed from consideration.
 * :attr:`grids_switch_thrd`: The threshold for the magnitude of the change in the density matrix that triggers the switch from the initial grid specified by :attr:`grids_level_i` to the final grid specified by :attr:`grids_level_f`.
-* :attr:`blockdim`: The maximum number of grid points to loop over at once. The number of grid points per batch is the minimum of :attr:`blockmin` and the maximum number of points allowed by the memory available for the calculation. The maximum memory can be adjusted by setting the :attr:`max_memory` attribute, which is initially set to :attr:`mol.max_memory`, the max memory setting for the Mole object.
+* :attr:`blockdim`: The maximum number of grid points to loop over at once. The number of grid points per batch is the minimum of :attr:`blockdim` and the maximum number of points allowed by the memory available for the calculation. The maximum memory can be adjusted by setting the :attr:`max_memory` attribute, which is initially set to :attr:`mol.max_memory`, the max memory setting for the Mole object.
+* :attr:`dfj`: If :attr:`True`, density fitting is used for the J-matrix. If :attr:`False`, SGX is used.
+* :attr:`direct_j`: If :attr:`True`, direct evaluation of the J-matrix is used (slow, for testing only).
+* :attr:`pjs`: If :attr:`True`, P-junction screening is used. Threshold determined by :attr:`direct_scf_tol`.
+
+:attr:`SGXHF` attribute:
+
+* :attr:`direct_scf_sgx`: Whether to use direct SCF within the SGX module, meaning that the J and K matrices are evaluated from the difference in the density matrix from the previous iteration.
```

Comment:
Incremental Fock builds typically also employ resets i.e. full Fock builds every N iterations. How is N controlled in your implementation?

If you don't do full builds, you suffer from creeping in numerical error.

Reply:
I'm not currently doing rebuilds except for when the grid level changes, and I did observe that this causes small numerical noise even for simple closed-shell organic systems (~1e-7 Eh). The direct JK evaluation in PySCF does not do rebuilds either, though I have not noticed it causing as much noise as the direct SGX version. I can add this if you want it (should be a fairly simple change), though I don't have any experience with what values of N are sensible.

Reply:
I think most programs do full rebuilds every 5 iterations or so.